In [36]:
import numpy as np
import pandas as pd

In [37]:
np.__version__

'1.21.2'

In [38]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'

In [39]:
# !wget data -o cars_data.csv

In [40]:
df = pd.read_csv('cars_data.csv')

In [41]:
# to get number of records, I just check the shape
df.shape

(11914, 16)

The number of records are: **11914** in number

In [42]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [43]:
# find the top 3
df['Make'].value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

Top 3 from the above is: **Chevrolet, Ford, Volkswagen**

In [44]:
# find the unique values for Audi Make
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [45]:
(df['Make'] == 'Audi').nunique()

2

In [46]:
# filter dataset to check if the make is equals "Audi" 
# and then display all the unique properties of the Audi car

df[df['Make'].str.contains('Audi')].nunique()

Make                   1
Model                 34
Year                  20
Engine Fuel Type       5
Engine HP             40
Engine Cylinders       6
Transmission Type      3
Driven_Wheels          2
Number of Doors        2
Market Category       15
Vehicle Size           3
Vehicle Style          6
highway MPG           23
city mpg              18
Popularity             1
MSRP                 234
dtype: int64

We have **34 models** in the Audi car

In [47]:
# number of empty rows
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

We have **5 columns** with empty values

In [48]:
df['Engine Cylinders'].describe()

count    11884.000000
mean         5.628829
std          1.780559
min          0.000000
25%          4.000000
50%          6.000000
75%          6.000000
max         16.000000
Name: Engine Cylinders, dtype: float64

In [49]:
df['Engine Cylinders'].median()

6.0

The 50% percentile is the same as median which is **6.00**

In [50]:
df['Engine Cylinders'].mode()

0    4.0
dtype: float64

In [51]:
# fill the null values with mode value
df_engine = df['Engine Cylinders'].fillna(4.0)

In [52]:
df_engine.tail(15)

11899    6.0
11900    6.0
11901    4.0
11902    6.0
11903    8.0
11904    8.0
11905    8.0
11906    6.0
11907    6.0
11908    6.0
11909    6.0
11910    6.0
11911    6.0
11912    6.0
11913    6.0
Name: Engine Cylinders, dtype: float64

In [53]:
df_engine.median()

6.0

No the MEAN **did not change**

In [54]:
# select lotus car
df_lotus = df[df['Make'] == 'Lotus']

In [55]:
df_lotus.head()

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   
3915  Exotic,High-Performance      Compact   Convertible           27   
3916  Exotic,High-Performance      Compact   Convertible           26   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250  
3915        21         613  47250  
3916        20         613  54990

In [56]:
# select with only engine hp and engine cylinders
df_lotus_with_engine_hp_and_engine_cylinders = df_lotus[['Engine HP', 'Engine Cylinders']]
df_lotus_with_engine_hp_and_engine_cylinders.reset_index(drop=True).head()

Engine HP  Engine Cylinders
0      189.0               4.0
1      218.0               4.0
2      189.0               4.0
3      189.0               4.0
4      218.0               4.0

In [57]:
df_lotus_with_engine_hp_and_engine_cylinders = df_lotus_with_engine_hp_and_engine_cylinders.drop_duplicates()

In [58]:
df_lotus_with_engine_hp_and_engine_cylinders.shape

(9, 2)

It is 9 rows as expected

In [59]:
X = np.array(df_lotus_with_engine_hp_and_engine_cylinders)

In [60]:
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [61]:
# matrix-mulitplication
XTX = X.T.dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [62]:
# inverse is:
X_inv = np.linalg.inv(XTX)

In [63]:
X_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [64]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [65]:
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [66]:
product_X = X_inv.dot(X.T)

In [67]:
w = product_X.dot(y)

In [68]:
w

array([  4.59494481, -63.56432501])

The value of the first element in *w* is **4.59494481**